core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -221,18 +220,16 @@ public long getLastComputedLeftDelayNanos() {
 
     /**
      * Updates delay left based on current time (in nanoseconds) and index/node settings.
-     * Should only be called from ReplicaShardAllocator.
      * @return updated delay in nanoseconds
      */
     public long updateDelay(long nanoTimeNow, Settings settings, Settings indexSettings) {
-        long delayTimeoutMillis = getAllocationDelayTimeoutSetting(settings, indexSettings);
+        long delayTimeoutNanos = getAllocationDelayTimeoutSettingNanos(settings, indexSettings);
         final long newComputedLeftDelayNanos;
-        if (delayTimeoutMillis == 0l) {
+        if (delayTimeoutNanos == 0l) {
             newComputedLeftDelayNanos = 0l;
         } else {
             assert nanoTimeNow >= unassignedTimeNanos;
-            long delayTimeoutNanos = TimeUnit.NANOSECONDS.convert(delayTimeoutMillis, TimeUnit.MILLISECONDS);
-            newComputedLeftDelayNanos = Math.max(0l, delayTimeoutNanos - (nanoTimeNow - unassignedTimeNanos));
+            newComputedLeftDelayNanos = delayTimeoutNanos - (nanoTimeNow - unassignedTimeNanos);
```

Comment:
This can be negative. This breaks consistency as the documentation for `lastComputedLeftDelayNanos` says `// how long to delay shard allocation, not serialized (always positive, 0 means no delay)`.